Mathlib/GroupTheory/SpecificGroups/HNNExtension.lean

```diff
@@ -0,0 +1,655 @@
+/-
+Copyright (c) 2023 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes
+-/
+
+import Mathlib.GroupTheory.SpecificGroups.Coprod
+import Mathlib.GroupTheory.Complement
+
+/-!
+
+## HNN Extensions of Groups
+
+This file defines the HNN extensions of a group `G`, `HNNExtension G A B φ`. Given a group `G`,
+subgroups `A` and `B` and an isomorphism `φ` of `A` and `B`, we adjoin a letter `t` to `G`, such
+that for any `a ∈ A`, the conjugate of `of a` by `t` is `of (φ a)`, where `of` is the canoncial map
+from `G` into the `HNNExtension`.
```

Comment:
I guess this uniquely describes HNN extensions. However, I've never heard of them before. Could you please add a bit more to this documentation? A reference to mainstream literature? And possible one or two main applications of HNN extensions? If Britton's lemma is the main application, please already mention it in this free-form text. But it would also be nice to know where Britton's lemma comes up. (Because I hadn't heard of that before either.)

Reply:
I always pick things out of textbooks to do in Lean I don't fully understand myself. I'm currently working towards Freiheitsatz, which says certain subgroups of groups presented by a single relation are free. Britton's lemma is I guess a lemma that proves the relation you have to quotient by to form the HNN extension is small. I guess the use of HNN extensions and Brittons lemma, is that you can prove other groups (probably a quotient group) either embed into or are isomorphic to an HNN extension, and then prove results about other groups. It's got some algebraic topology analogue as well, there's some topological construction that forms a space whose fundamental group is the HNN Extension.

Are you happy for an incomplete description by a non-expert? I think @kmill might know more

Reply:
HNN extensions are studied by low-dimensional topologists and geometric group theorists since they're very common. Whereas the van Kampen theorem gives amalgamated free products, HNN extensions are what you need to handle loops in what you're gluing together. Bundles over a circle have fundamental groups that are HNN extensions (conjugation by t is the "monodromy action"), and knot complements also have such fundamental groups -- every compact oriented 3-manifold with a non-spherical boundary component has one.

Britton's lemma is far from the main application. It's just an important structural result about words.

(Typo in "canonical" by the way.)

Reply:
@kmill Thanks for chiming in! Since you seem to know this bit of math. Would you mind reviewing this PR and/or maybe suggesting something for the module docstring.